### PR TITLE
feat: CosmWasm custom msg parsers and queriers for modules

### PR DESCRIPTION
**Pull request by leobragaz**
## Description

This PR add: 
- `x/profiles`
- `x/subspaces`
- `x/posts`
- `x/relationships`
- `x/reports`
- `x/reactions`

cosmWasm message parsers and queriers in order to allow interactions with Desmos chain from within smart contracts.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://docs.cosmos.network/v0.44/building-modules/intro.html)
- [x] included the necessary unit and integration [tests](https://github.com/desmos-labs/desmos/blob/master/CONTRIBUTING.md#testing)
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: app/sim-params.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"op_reflect_contract_path":"../testutil/wasm/reflect.wasm"}
```
Comment (dadamu) [inline app/sim-params.json:1]: This file can be removed.

File: cosmwasm/msg_parser.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+type CustomMsg struct {
+	Route   string          `json:"route"`
+	MsgData json.RawMessage `json:"msg_data"`
+}
+
+func (router ParserRouter) ParseCustom(contractAddr sdk.AccAddress, data json.RawMessage) ([]sdk.Msg, error) {
+	var customMsg CustomMsg
+	err := json.Unmarshal(data, &customMsg)
+
+	log.Println("[!] CosmWasm contract msg routed to module: ", customMsg.Route)
```
Comment (dadamu) [inline cosmwasm/msg_parser.go:52]: The message is for debug, it would be better to be `log.Debug`, instead.

File: cosmwasm/query_parser.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	QueryRoutePosts         = poststypes.ModuleName
+	QueryRouteReports       = reportstypes.ModuleName
+	QueryRouteReactions     = reactionstypes.ModuleName
+)
+
+func (q QuerierRouter) QueryCustom(ctx sdk.Context, data json.RawMessage) ([]byte, error) {
+	var customQuery CustomQuery
+	err := json.Unmarshal(data, &customQuery)
+
+	log.Println("[!] Cosmwasm contract query routed to module: ", customQuery.Route)
```
Comment (dadamu) [inline cosmwasm/query_parser.go:52]: As above.

File: cosmwasm/msg_parser.go (hunk 2 above)
Comment (bragaz) [reply cosmwasm/msg_parser.go:52]: The `Debug` function doesn't exist for the `log` package.

File: cosmwasm/msg_parser.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	WasmMsgParserRouteProfiles      = profilestypes.ModuleName
+	WasmMsgParserRouteSubspaces     = subspacestypes.ModuleName
+	WasmMsgParserRouteRelationships = relationshipstypes.ModuleName
+	WasmMsgParserRoutePosts         = poststypes.ModuleName
+	WasmMsgParserRouteReports       = reportstypes.ModuleName
+	WasmMsgParserRouteReactions     = reactionstypes.ModuleName
+)
+
+type MsgParserInterface interface {
+	Parse(contractAddr sdk.AccAddress, msg wasmvmtypes.CosmosMsg) ([]sdk.Msg, error)
```
Comment (dadamu) [inline cosmwasm/msg_parser.go:29]: Can we remove this method? It seems unused in the module wasm parser.

File: cosmwasm/query_parser.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	poststypes "github.com/desmos-labs/desmos/v3/x/posts/types"
+	profiletypes "github.com/desmos-labs/desmos/v3/x/profiles/types"
+	reactionstypes "github.com/desmos-labs/desmos/v3/x/reactions/types"
+	relationshipstypes "github.com/desmos-labs/desmos/v3/x/relationships/types"
+	reportstypes "github.com/desmos-labs/desmos/v3/x/reports/types"
+	subspacestypes "github.com/desmos-labs/desmos/v3/x/subspaces/types"
+)
+
+type Querier interface {
+	Query(ctx sdk.Context, request wasmvmtypes.QueryRequest) ([]byte, error)
```
Comment (dadamu) [inline cosmwasm/query_parser.go:20]: As above

File: cosmwasm/msg_parser.go (hunk 2 above)
Comment (dadamu) [reply cosmwasm/msg_parser.go:52]: Okay, can we remove it instead?

File: x/subspaces/keeper/common_test.go [hunk 6]
```diff
@@ -30,6 +30,10 @@ type KeeperTestsuite struct {
 	storeKey       sdk.StoreKey
 }
 
+func TestTestSuite(t *testing.T) {
```
Comment (dadamu) [inline x/subspaces/keeper/common_test.go:33]: Why change it?

File: cosmwasm/msg_parser.go (hunk 2 above)
Comment (bragaz) [reply cosmwasm/msg_parser.go:52]: Yes I think so, it was for test purposes anyway

**Review by manu0466 — APPROVED** [review]
Looks good to me

File: x/subspaces/keeper/common_test.go (hunk 6 above)
Comment (bragaz) [reply x/subspaces/keeper/common_test.go:33]: I guess an error

**Comment by dadamu** [comment]
@bragaze I've improved the naming and function structures to be same format.

File: x/profiles/wasm/querier_test.go [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+							"dtag",
+							"cosmos10nsdxxdvy9qka3zv0lzw8z9cnu6kanld8jh773",
+							"cosmos19xz3mrvzvp9ymgmudhpukucg6668l5haakh04x",
+						),
+					},
+					Pagination: &query.PageResponse{NextKey: nil, Total: 1}},
+			),
+		},
+		{
+			name: "chain links request request is parsed correctly",
```
Comment (dadamu) [inline x/profiles/wasm/querier_test.go:84]:
```suggestion
			name: "chain links request is parsed correctly",
```

File: x/profiles/wasm/querier_test.go [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+							),
+							types.NewChainConfig("cosmos"),
+							time.Date(2020, 1, 1, 00, 00, 00, 000, time.UTC),
+						),
+					},
+					Pagination: &query.PageResponse{NextKey: nil, Total: 1}},
+			),
+		},
+		{
+			name: "chain link owners request request is parsed correctly",
```
Comment (dadamu) [inline x/profiles/wasm/querier_test.go:125]:
```suggestion
			name: "chain link owners request is parsed correctly",
```

File: x/reports/wasm/msg_parser_test.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+				"cosmos1qzskhrcjnkdz2ln4yeafzsdwht8ch08j4wed69",
+			)),
+			shouldErr: false,
+			expMsgs: []sdk.Msg{types.NewMsgDeleteReport(
+				1,
+				1,
+				"cosmos1qzskhrcjnkdz2ln4yeafzsdwht8ch08j4wed69")},
+		},
+		{
+			name: "support standardReason message is parsed correctly",
```
Comment (dadamu) [inline x/reports/wasm/msg_parser_test.go:72]:
```suggestion
			name: "support standard reason message is parsed correctly",
```

File: x/reports/wasm/querier_test.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+		expResponse []byte
+	}{
+		{
+			name:        "wrong request type returns error",
+			request:     wrongQueryBz,
+			shouldErr:   true,
+			expResponse: nil,
+		},
+		{
+			name:    "reports request is parsed correctly",
```
Comment (dadamu) [inline x/reports/wasm/querier_test.go:36]:
```suggestion
			name:    "reports query request is parsed correctly",
```

File: x/subspaces/wasm/msg_parser_test.go [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+		expMsgs   []sdk.Msg
+	}{
+		{
+			name:      "parse wrong module message returns error",
+			msg:       wrongMsgBz,
+			shouldErr: true,
+			expMsgs:   nil,
+		},
+		{
+			name: "create subspace json message is parsed correctly",
```
Comment (dadamu) [inline x/subspaces/wasm/msg_parser_test.go:37]:
```suggestion
			name: "create subspace message is parsed correctly",
```

File: x/subspaces/wasm/msg_parser_test.go [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+			expMsgs: []sdk.Msg{types.NewMsgCreateSubspace(
+				"test",
+				"test",
+				"cosmos1qzskhrcjnkdz2ln4yeafzsdwht8ch08j4wed69",
+				"cosmos1qzskhrcjnkdz2ln4yeafzsdwht8ch08j4wed69",
+				"cosmos1qzskhrcjnkdz2ln4yeafzsdwht8ch08j4wed69",
+			)},
+		},
+		{
+			name: "edit subspace json message is parsed correctly",
```
Comment (dadamu) [inline x/subspaces/wasm/msg_parser_test.go:57]:
```suggestion
			name: "edit subspace message is parsed correctly",
```

File: x/subspaces/wasm/querier_test.go [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+		expResponse []byte
+	}{
+		{
+			name:        "wrong request type returns error",
+			request:     wrongQueryBz,
+			shouldErr:   true,
+			expResponse: nil,
+		},
+		{
+			name:    "subspaces request is parsed correctly",
```
Comment (dadamu) [inline x/subspaces/wasm/querier_test.go:36]:
```suggestion
			name:    "subspaces query request is parsed correctly",
```

File: x/subspaces/wasm/querier_test.go [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+							"cosmos1m0czrla04f7rp3zg7dsgc4kla54q7pc4xt00l5",
+							time.Date(2020, 1, 1, 12, 00, 00, 000, time.UTC),
+						),
+					},
+					Pagination: &query.PageResponse{NextKey: nil, Total: 1},
+				},
+			),
+		},
+		{
+			name:    "subspace request is parsed correctly",
```
Comment (dadamu) [inline x/subspaces/wasm/querier_test.go:68]:
```suggestion
			name:    "subspace query request is parsed correctly",
```

**Review by dadamu — CHANGES_REQUESTED** [review]
Overall looks good to me, just having some test case name issue.

**Review by dadamu — APPROVED** [review]
Ready to ship to me

**Pull request merged**